monty-31/src/aarch64_neon/poseidon2.rs

```diff
@@ -93,9 +340,206 @@ where
     fn permute_state_terminal(&self, state: &mut [PackedMontyField31Neon<FP>; WIDTH]) {
         external_terminal_permute_state(
             state,
-            self.external_constants.get_terminal_constants(),
-            add_rc_and_sbox_generic,
+            &self.packed_terminal_external_constants,
+            add_rc_and_sbox::<FP, D>,
             &MDSMat4,
         );
     }
 }
+
+/// Converts a scalar constant into a packed NEON vector in "negative form".
+///
+/// This is a pre-computation step for round constants. Instead of storing a constant `c`
+/// and performing a modular addition `(val + c) mod P` in the permutation loop, we pre-compute
+/// `c' = c - P`. This allows the round update to be a simple, non-modular signed integer
+/// addition `val + c'`, which is significantly faster.
+#[inline(always)]
+fn convert_to_vec_neg_form_neon<MP: MontyParameters>(input: i32) -> uint32x4_t {
+    unsafe {
+        // Perform the subtraction as a standard integer operation.
+        //
+        // This computes the "negative form" of the constant.
+        let input_sub_p = input - (MP::PRIME as i32);
+
+        // Broadcast (duplicate) the scalar result into all four lanes of a 128-bit NEON vector.
+        let vec_s32 = aarch64::vdupq_n_s32(input_sub_p);
+
+        // Reinterpret the bits of the signed vector as an unsigned vector.
+        //
+        // This is a zero-cost operation that only changes the type for the Rust compiler.
+        aarch64::vreinterpretq_u32_s32(vec_s32)
+    }
+}
+
+/// Performs the AddRoundConstant and S-Box operations of a Poseidon round (`x -> (x + c)^D`).
+///
+/// This function combines two steps for efficiency. It takes a vector of field elements `val`
+/// and adds the pre-formatted round constant `rc`. The result is then raised to the power `D`.
+///
+/// A key optimization is that the round constants `rc` are stored in a negative form `c - P`.
+/// This allows for a simple integer addition `val + rc` without an intermediate modular
+/// reduction. The result of this addition is in the range `[-P, P)`. This non-canonical
+/// result is then reduced back to `[0, P)` before the exponentiation is applied.
+///
+/// # Safety
+/// - `val` must contain elements in canonical form `[0, P)`.
+/// - `rc` must contain elements represented as `c - P`, which are in `[-P, 0]`.
+#[inline(always)]
+fn add_rc_and_sbox<PMP, const D: u64>(val: &mut PackedMontyField31Neon<PMP>, rc: uint32x4_t)
```

Comment:
For now, just have this accept a monty field element.
```suggestion
fn add_rc_and_sbox<PMP, const D: u64>(val: &mut PackedMontyField31Neon<PMP>, rc: MontyField31<PMP>)
```